ITR/data/base_providers.py

```diff
@@ -358,44 +392,63 @@ def get_company_intensity_and_production_at_base_year(self, company_ids: List[st
-            [ICompanyData.parse_obj(c.dict()).dict() for c in self.get_company_data(company_ids)],
-            exclude=['projected_targets', 'projected_intensities', 'historic_data']).set_index(
-                self.column_config.COMPANY_ID)
+        excluded_cols = ['projected_targets', 'projected_intensities', 'historic_data']
+        df = pd.DataFrame.from_records(
+            [ICompanyData.parse_obj({k:v for k, v in c.dict().items() if k not in excluded_cols}).dict()
+             for c in self.get_company_data(company_ids)]).set_index(self.column_config.COMPANY_ID)
+        return df
 
-    def get_company_projected_trajectories(self, company_ids: List[str]) -> pd.DataFrame:
+    def get_company_projected_trajectories(self, company_ids: List[str], year=None) -> pd.DataFrame:
         """
         :param company_ids: A list of company IDs
-        :return: A pandas DataFrame with projected intensity trajectories per company, indexed by company_id
-        """
-        trajectory_list = [self._convert_projections_to_series(c, self.column_config.PROJECTED_EI, EScope.S1S2)
-                           for c in self.get_company_data(company_ids)]
-        if trajectory_list:
-            with warnings.catch_warnings():
-                # pd.DataFrame.__init__ (in pandas/core/frame.py) ignores the beautiful dtype information adorning the pd.Series list elements we are providing.  Sad!
-                warnings.simplefilter("ignore")
-                return pd.DataFrame(trajectory_list)
+        :param year: values for a specific year, or all years if None
+        :return: A pandas DataFrame with projected intensity trajectories per company, indexed by company_id and scope
+        """
+        projected_trajectories = [(c.company_id, scope, getattr(c.projected_intensities, scope.name).projections)
+                                  # FIXME: we should make _companies a dict so we can look things up rather than searching every time!
+                                  for c in self._companies for scope in EScope.get_result_scopes()
+                                  if c.company_id in company_ids
+                                  if getattr(c.projected_intensities, scope.name)]
+        if projected_trajectories:
+            if year:
+                company_ids, scopes, values = list(map(list, zip(*[(pt[0], pt[1], [yvp.value for yvp in pt[2] if yvp.year==year][0]) for pt in projected_trajectories])))
+                index=pd.MultiIndex.from_tuples(zip(company_ids, scopes), names=["company_id", "scope"])
+                return pd.Series(data=values, index=index)
+            else:
+                company_ids, scopes, yvp_dicts = list(map(list, zip(*[(pt[0], pt[1], {yvp.year:yvp.value for yvp in pt[2]}) for pt in projected_trajectories])))
+                index=pd.MultiIndex.from_tuples(zip(company_ids, scopes), names=["company_id", "scope"])
+                return pd.DataFrame(data=yvp_dicts, index=index)
```

Comment:
sometimes we return DataFrame, and sometimes - Series.
does it work?

Reply:
It does...and respects the fact that Pandas itself is quite polymorphic when it comes to these two.  I submitted an issue complaining that the documentation for a Series function was described only in terms of DataFrame (or vice-versa) and they said that any functions that operate on both types are supposed to be the same as much as possible.  If the caller knows they are just asking for a year of data, they know they'll get back the year column--which is a series.  If not asking for a specific year, they'll get back the whole dataframe.  Nobody wants to get back a dataframe with only one column that they must then re-index just to get the values they want.

Reply:
Shall we reflect this poly-return in function definition and function comment?

Reply:
I don't know a good way to do this, and Pandas, which uses annotations and typing extensively (but not pydantic), doesn't make this distinction.  Here's some text from the Pandas merge function (https://github.com/pandas-dev/pandas/blob/v1.5.2/pandas/core/frame.py#L10071-L10104):

```
_merge_doc = """
Merge DataFrame or named Series objects with a database-style join.
A named Series object is treated as a DataFrame with a single named column.
```

I believe we are following the Pandas patterns correctly here and there's no issue.

Reply:
but now we need to insert `if yeasr is None` here and there, because it can be either a Dataframe or a Series.
If we always returned a Dataframe - other functions would be able to become more linear, too.

instead of Series, we could return 1-dimentional Dataframe - and then post-processing could be more straightforward.
That's how it was before this PR, right?

Reply:
I've read a lot of Pandas code in the past 2 years.  I used to think that analysis about the 1-dimensional dataframe was correct.  I have come to learn that it's just not the Pandas way of doing things.  Everything "just works" when 1-dimensional arrays are pd.Series, and it actually behaves more naturally.  The only thing that's confusing is that typing annotations don't seem to grok this very well.  But everybody has learned to ignore that detail.  I'm happy to be proved wrong, but that's what I've learned so far.